timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/HistoryAnalyzer.java

```diff
@@ -195,4 +206,35 @@ private static int getQuorumSize(List<ConsolidatedLearnerAndAcceptorRecord> reco
     private static byte[] getPaxosValueData(Optional<PaxosValue> learnedValue) {
         return learnedValue.map(PaxosValue::getData).orElse(null);
     }
+
+    private static boolean clockWentBackwardsOnNode(ConsolidatedLearnerAndAcceptorRecord record) {
+        List<SequenceAndTimestampPair> seqAndTimestampPairs = KeyedStream.stream(record.record())
+                .map(val -> getPaxosValueData(val.learnedValue()))
+                .filter(Predicates.notNull())
+                .mapEntries((sequence, timestamp) -> Maps.immutableEntry(
+                        sequence, ImmutableSequenceAndTimestampPair.of(sequence, PtBytes.toLong(timestamp))))
+                .values()
+                .sorted(Comparator.comparingLong(SequenceAndTimestampPair::sequence))
+                .collect(Collectors.toList());
+
+        long greatestTimestampSoFar = -1L;
+        long currentTimestamp;
+        for (SequenceAndTimestampPair sequenceAndTimestampPair : seqAndTimestampPairs) {
+            currentTimestamp = sequenceAndTimestampPair.timestamp();
+            if (currentTimestamp <= greatestTimestampSoFar) {
+                return true;
+            }
+            greatestTimestampSoFar = currentTimestamp;
+        }
+        return false;
+    }
+
+    @Value.Immutable
+    interface SequenceAndTimestampPair {
+        @Value.Parameter
+        long sequence();
+
+        @Value.Parameter
+        long timestamp();
+    }
```

Comment:
If you want to be a bit fancy:
```suggestion
        Stream<Long> expectedSortedTimestamps = KeyedStream.stream(record.record())
                .map(val -> getPaxosValueData(val.learnedValue()))
                .filter(Predicates.notNull())
                .mapEntries((sequence, timestamp) -> Maps.immutableEntry(sequence, PtBytes.toLong(timestamp)))
                .entries()
                .sorted(Comparator.comparingLong(Entry::getKey))
                .map(Entry::getValue);
        return StreamEx.of(expectedSortedTimestamps)
                .pairMap((first, second) -> first >= second)
                .anyMatch(x -> x);
```
If you don't want to use `StreamEx`, you can still avoid the immutable by checking `expectedSortedTimestamps` in a for loop, but just compare neighbours instead of maintaining a global maximum